### DISCUSSION TITLE
List all with id as unique partition key?

### DISCUSSION THREAD

**Discussion by brandonpsmith** [tywalch/electrodb#210] [Q&A]
I'm trying to query all `locations`. How do I query begins with `LOCATION#` and get all `locations`? ...

```js
{
    model: {
      version: "1",
      entity: "location",
      service: "service",
    },
    attributes: {
      locationId: {
        type: "string",
        required: true,
        readOnly: true,
      },
      name: {
        type: "string",
        required: true,
      },
      coordinate: {
        type: "map",
        required: true,
        properties: {
          latitude: {
            type: "string",
            required: true,
          },
          longitude: {
            type: "string",
            required: true,
          },
        },
      },
      address: {
        type: "map",
        properties: {
          street: {
            type: "string",
          },
          city: {
            type: "string",
          },
          state: {
            type: "string",
          },
          zip: {
            type: "string",
          },
          county: {
            type: "string",
          },
        },
      },
    },
    indexes: {
      primary: {
        pk: {
          field: "pk",
          casing: "upper",
          template: "LOCATION#${locationId}",
          composite: ["locationId"],
        },
        sk: {
          field: "sk",
          casing: "upper",
          template: "LAT#${coordinate.latitude}#LON#${coordinate.longitude}",
          composite: ["coordinate.latitude", "coordinate.longitude"],
        },
      },
    },
  }


```

```ts
static async list() {
    const { data } = await LocationEntity.query.primary({ ???? }).begins({ ???? }).go();
    return data;
  }
```

**Comment by tywalch** [comment]
Hi @brandonpsmith 👋 

With ElectroDB, you only have to interact with the individual PK and SK composite attributes themselves -- how you provide these to ElectroDB means different things:

```typescript
// Access queries on the `query` object
await LocationEntity.query
  // The "Access Pattern Method"
  // Must at a minimum include the attributes that make up the PK, but also accepts sort key attributes
  // If you provide a sort key attribute, ElectroDB will treat this value as being complete (e.g. The "complete" for the location "Texas", not "Tex")
  // Lastly, you still must provide your sort keys in the order they are defined in your "composite". ElectroDB is building a string after all, and it can't 
  // build a key out of left-to-right order.
  .primary({ location: 'Georgia', latitude: '33.753746' }) 
  // The "Sort Key Condition Methods" allow for partial sort keys or unique conditions on sort keys (e.g. 'between', 'begins', 'gte', 'gt', 'lte', 'lt')
  .begins({ Longitude: '-8' })
  // complete the query
  .go()
```

Let me know if this helps, you can find more detail on this here: https://electrodb.dev/en/queries/query/

**Reply by brandonpsmith** [reply]
It helps but doesn't solve listing all locations with zero params. Is there a way to query this or is this just a scan? In this case. I don't have a state or coordinate to start with. I just need to query all locations.

**Reply by ikushlianski** [reply]
I saw one guy posting [this playground](https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2AT0IBlTADcsAY1IBfbhiiSwAIkRwjVSzABGNgNzijUPrgVUGFwRSAF5bTShUDAB9AKC4aL4GSTgPLx8-MAAFZ1RHNQ0qTTAwvjgWVXUtAAowUTAleoawSVhEAC5G5ub9TFwsH06bAEYbfCbu-kLNIZzYPKoxiebiDEMTIbl4DFwl7tNx7oYqJywXVCo8TsVlhpYmODpI4gBJGGvb5qLpOCG-bD4AHM9t0wAdPtJco43h9QV9ND8-qcgSC4eQ0FhyO8wE50IdQeC4UkUrC4TiEb9bP8sCj8Wi4BisZ1cXA6c1CaD4LgjNhpBpBl0yd9KTZqbTPhzupDjJSbmTyYjbHxUJIXJhUey2Q1vKh1KS4cKhsrVeqtWC2ZKwDT4AAPK6C7rAYA4uhYIQIN00MCQ+aOXCfBpOq1UADkQikPwwXAcCnOCigAGsGJoIVCaPrQdIExm4VwcAhsTYsxq4d4ZFB+pdOgBtAC6Zoalu6uGzDrJecQhZbJdBZdklcp1aLabeNnrAbBErN93IT1QxDmMAWuBzzW8CGCjgGfCGM8ezzgi+XPYa1rgNqGgP6owb3tbcvl3HzhavWGGxdv2qs-fYg5se7nV4YDHW8mxWe8JwaDsC0va9u0-MA+wrX8a2HX0qFHcd5VMKdlhwxt8UUOJgnBABKdJfAUAB1B5ALgAotBKMAygqBiilqJoHwaVp4AQVdeh2bchhvT4pi0IYaNnA8e1WdYRS2PoNUtY5Tjje0uOaACDxhNtukNKlkWBW90VQTE4GxFkzTA4lZQnfTRUMk8yAZUymRxDA8VwuEGBgGByFwFddPhRUHIBIyJxMsyLI81kvNBSQGCSQFMFXBp7LFcK4UtS0zztQKNLuWiD3yicgw4N0wA9PwhC0+c8DqeUg3YMMrRkTBo0cMA4zARNkwnddN35HdbFq4hdlvLNUuaaDCw-SDEO-ZCqzAOtQNvFspqg58-gTJyv3LAdUNGuBMNAuLumOo8-SmgaYyEkairqq6aCc3LYLfJzJqC7oZve99doQpDDpW-9HqAkCJzAhoNu+6bttsV9hng+agZQlasLJfCCQtJoOSIwISNEcjxAYXBND4IxuF1LcfEQ8hjkPNNqhszoMsILkeSwPltwAflZwywAAH2YygEHxaUTD5kWTQwIWRY3QgdXUKXjTVWXheVDdSLbcgqCiPhsjTNjNAAOh5OAGeqArvRHLsrDgABaVBTOAwgnUoLAYGqUizRss0Od5IbOgDrmhqF4WbB7CXKWj8OwAABjNJWqE6ZO4-jvDSJNwEoG90QcJJsmKapimw-NhnJP3OqreY5JWTAHy-LwIRTGuWuSSIQzCEb-zcCljK5c1sWbfQt5+4FjXRbBbWH11-WwErujjZNppy8uK3lh75vOi3gK48jtk-eWY6dNFe2HagDhMCdl2xjAIMPa9n3caznO84LoMABVXRquQEyEQE9MqAAEJRCk3JpTLg1Mw6NwgIlBgyUMDVBPtiNm7dfgZRnssOeGADaLwPMvVA0gYCW2tigi0WdiBUA3nCBKSUUroIodnXOyh-KyF8CKBgG5EjlBsNPfO4ggwvBDNYKBG5igIArAoJ4FQqBQDADAeR5UVyiCEawRKlwYA4nkebC2lxCDdWEdYBgjhUBcIQMUIhJCNBAlUc6KijAFAvHUeoK0QhcD23KkCSqWAExwBNgEuxYB5w0kBC6UgPolzQhgIQKh4SwD6C4egIJUAuDqKknVN4rB2CcAYAtXUVAwFF0gdA7cDdfLPU-lAfB1cUH8zCoQSJCwdKYJ1nAPWuDDboUIcQ0hqZR4xMzibKhNDQTkM+EGZOnQADU0yhlv3IgXbwlF0lVyAkxEMLA6AO2GPHeOIZPDLMyE06JGziEOwAKx7IOSTCpaYqk1OIMgsGJ0YkjyiRhGApF3BAA) at `electrodb.fun`

Line 92. You could create a PK with an empty array, and some SK. So your PK will be generated as the name of service. So when querying that service you could get all the items with that PK, thus "listing" all the items belonging to that service.

**Reply by chrisfitkin** [reply]
Is that a good practice?  I saw a similar config in [this github example](https://github.com/serverless-stack/sst-weekly-repos/blob/main/2022-08-02-practical-graphql-with-pothos/services/core/article.ts).

```js
indexes: {
      primary: {
        pk: {
          field: "pk",
          composite: [],
        },
        sk: {
          field: "sk",
          composite: ["articleID"],
        },
      },
    },
```

And then accessed like this to fetch all.

```js
export async function list() {
  return ArticleEntity.query.primary({}).go();
}
```

**Reply by tywalch** [reply]
> I saw one guy posting [this playground](https://electrodb.fun/?#code/PQKgBAsg9gJgpgGzARwK5wE4Es4GcA0YuccYGeqCALgUQBYCG5YA7llXWAGbZwB2MXGBDAAUKKwBbAA5QMVMAG8wAUT5V2AT0IBlTADcsAY1IBfbhiiSwAIkRwjVSzABGNgNzijUPrgVUGFwRSAF5bTShUDAB9AKC4aL4GSTgPLx8-MAAFZ1RHNQ0qTTAwvjgWVXUtAAowUTAleoawSVhEAC5G5ub9TFwsH06bAEYbfCbu-kLNIZzYPKoxiebiDEMTIbl4DFwl7tNx7oYqJywXVCo8TsVlhpYmODpI4gBJGGvb5qLpOCG-bD4AHM9t0wAdPtJco43h9QV9ND8-qcgSC4eQ0FhyO8wE50IdQeC4UkUrC4TiEb9bP8sCj8Wi4BisZ1cXA6c1CaD4LgjNhpBpBl0yd9KTZqbTPhzupDjJSbmTyYjbHxUJIXJhUey2Q1vKh1KS4cKhsrVeqtWC2ZKwDT4AAPK6C7rAYA4uhYIQIN00MCQ+aOXCfBpOq1UADkQikPwwXAcCnOCigAGsGJoIVCaPrQdIExm4VwcAhsTYsxq4d4ZFB+pdOgBtAC6Zoalu6uGzDrJecQhZbJdBZdklcp1aLabeNnrAbBErN93IT1QxDmMAWuBzzW8CGCjgGfCGM8ezzgi+XPYa1rgNqGgP6owb3tbcvl3HzhavWGGxdv2qs-fYg5se7nV4YDHW8mxWe8JwaDsC0va9u0-MA+wrX8a2HX0qFHcd5VMKdlhwxt8UUOJgnBABKdJfAUAB1B5ALgAotBKMAygqBiilqJoHwaVp4AQVdeh2bchhvT4pi0IYaNnA8e1WdYRS2PoNUtY5Tjje0uOaACDxhNtukNKlkWBW90VQTE4GxFkzTA4lZQnfTRUMk8yAZUymRxDA8VwuEGBgGByFwFddPhRUHIBIyJxMsyLI81kvNBSQGCSQFMFXBp7LFcK4UtS0zztQKNLuWiD3yicgw4N0wA9PwhC0+c8DqeUg3YMMrRkTBo0cMA4zARNkwnddN35HdbFq4hdlvLNUuaaDCw-SDEO-ZCqzAOtQNvFspqg58-gTJyv3LAdUNGuBMNAuLumOo8-SmgaYyEkairqq6aCc3LYLfJzJqC7oZve99doQpDDpW-9HqAkCJzAhoNu+6bttsV9hng+agZQlasLJfCCQtJoOSIwISNEcjxAYXBND4IxuF1LcfEQ8hjkPNNqhszoMsILkeSwPltwAflZwywAAH2YygEHxaUTD5kWTQwIWRY3QgdXUKXjTVWXheVDdSLbcgqCiPhsjTNjNAAOh5OAGeqArvRHLsrDgABaVBTOAwgnUoLAYGqUizRss0Od5IbOgDrmhqF4WbB7CXKWj8OwAABjNJWqE6ZO4-jvDSJNwEoG90QcJJsmKapimw-NhnJP3OqreY5JWTAHy-LwIRTGuWuSSIQzCEb-zcCljK5c1sWbfQt5+4FjXRbBbWH11-WwErujjZNppy8uK3lh75vOi3gK48jtk-eWY6dNFe2HagDhMCdl2xjAIMPa9n3caznO84LoMABVXRquQEyEQE9MqAAEJRCk3JpTLg1Mw6NwgIlBgyUMDVBPtiNm7dfgZRnssOeGADaLwPMvVA0gYCW2tigi0WdiBUA3nCBKSUUroIodnXOyh-KyF8CKBgG5EjlBsNPfO4ggwvBDNYKBG5igIArAoJ4FQqBQDADAeR5UVyiCEawRKlwYA4nkebC2lxCDdWEdYBgjhUBcIQMUIhJCNBAlUc6KijAFAvHUeoK0QhcD23KkCSqWAExwBNgEuxYB5w0kBC6UgPolzQhgIQKh4SwD6C4egIJUAuDqKknVN4rB2CcAYAtXUVAwFF0gdA7cDdfLPU-lAfB1cUH8zCoQSJCwdKYJ1nAPWuDDboUIcQ0hqZR4xMzibKhNDQTkM+EGZOnQADU0yhlv3IgXbwlF0lVyAkxEMLA6AO2GPHeOIZPDLMyE06JGziEOwAKx7IOSTCpaYqk1OIMgsGJ0YkjyiRhGApF3BAA) at `electrodb.fun`
> 
> 
> 
> Line 92. You could create a PK with an empty array, and some SK. So your PK will be generated as the name of service. So when querying that service you could get all the items with that PK, thus "listing" all the items belonging to that service. 
> 
> 

It actually is still tied to the entity through the sort key, so would be "listing" all items that belong to that entity 👍

**Reply by tywalch** [reply]
> Is that a good practice?  I saw a similar config in [this github example](https://github.com/serverless-stack/sst-weekly-repos/blob/main/2022-08-02-practical-graphql-with-pothos/services/core/article.ts).
> 
> 
> 
> ```js
> 
> indexes: {
> 
>       primary: {
> 
>         pk: {
> 
>           field: "pk",
> 
>           composite: [],
> 
>         },
> 
>         sk: {
> 
>           field: "sk",
> 
>           composite: ["articleID"],
> 
>         },
> 
>       },
> 
>     },
> 
> ```
> 
> 
> 
> And then accessed like this to fetch all.
> 
> 
> 
> ```js
> 
> export async function list() {
> 
>   return ArticleEntity.query.primary({}).go();
> 
> }
> 
> ```
> 
> 
> 
> 

It's definitely supported by ElectroDB 👍 Whether or not it makes sense to do with your data in particular depends on the cardinality of what you assign to the sort key. Hot partitions can still cause throughput problems. That said for something like "accountId", or something that has known bounds, it can be super useful to use this pattern. It allows me to create an initial query (for example, get all accountIds) and then use those values to seed a second round of queries.
